server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java

```diff
@@ -493,6 +495,16 @@ private ClusterState applyCreateIndexRequestWithV2Template(final ClusterState cu
 
         final Map<String, Object> mappings = resolveV2Mappings(request.mappings(), currentState, templateName, xContentRegistry);
 
+        if (request.dataStreamName() != null) {
+            DataStream dataStream = currentState.metadata().dataStreams().get(request.dataStreamName());
+            if (dataStream != null) {
+                String mappingPath = convertFieldPathToMappingPath(dataStream.getTimeStampField().getFieldName());
+                String parentObjectFieldPath = mappingPath.substring(0, mappingPath.lastIndexOf('.'));
+                Map<String, Object> parentObjectMapper = ObjectPath.eval("_doc." + parentObjectFieldPath, mappings);
+                parentObjectMapper.put(dataStream.getTimeStampField().getFieldName(), dataStream.getTimeStampField().getFieldMapping());
```

Comment:
Maybe assert that the `parentObjectMapper` does not contain a definition for the timestamp field already?

Should potentially be a production code check in case a component template can have the timestamp field too? Not sure if we guard against that.

Reply:
Currently the definition is coming from a composable index / component template. So if the name of the timestamp field mapping doesn't change then parentObjectMapper should always contain the mapping definition for the timestamp field.

Only if the name of the timestamp field changes then this logic will make sure that the original mapping definition will be used.

I think verifying whether no mapping definition exists, makes more sense when do this: https://github.com/elastic/elasticsearch/pull/58096#discussion_r440707325

In that case it is expected that the mapping definition for the timestamp_field originates not from composable index or component templates.

Reply:
Ah yes, sorry, I was one thought too far ahead.